pandas/core/dtypes/concat.py

```diff
@@ -21,11 +26,78 @@
 from pandas.core.construction import array, ensure_wrapped_if_datetimelike
 
 
+class NullArrayProxy:
+    """
+    Proxy object for an all-NA array.
+
+    Only stores the length of the array, and not the dtype. The dtype
+    will only be known when actually concatenating (after determining the
+    common dtype, for which this proxy is ignored).
+    Using this object avoids that the internals/concat.py needs to determine
+    the proper dtype and array type.
+    """
+
+    ndim = 1
+
+    def __init__(self, n: int):
+        self.n = n
+
+    @property
+    def shape(self):
+        return (self.n,)
+
+
+def _array_from_proxy(arr, dtype: DtypeObj, fill_value=lib.no_default):
+    """
+    Helper function to create the actual all-NA array from the NullArrayProxy object.
+
+    Parameters
+    ----------
+    arr : NullArrayProxy
+    dtype : the dtype for the resulting array
+    fill_value : scalar NA-like value
+        By default uses the ExtensionDtype's na_value or np.nan. For numpy
+        arrays, this can be overridden to be something else (eg None).
+
+    Returns
+    -------
+    np.ndarray or ExtensionArray
+    """
+    if is_extension_array_dtype(dtype):
+        return dtype.construct_array_type()._from_sequence(
+            [dtype.na_value] * arr.n, dtype=dtype
+        )
+    elif is_datetime64_ns_dtype(dtype):
+        from pandas.core.arrays import DatetimeArray
+
+        return DatetimeArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    elif is_timedelta64_ns_dtype(dtype):
+        from pandas.core.arrays import TimedeltaArray
+
+        return TimedeltaArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    else:
+        if is_integer_dtype(dtype):
+            dtype = "float64"
+            fill_value = np.nan
+        elif is_bool_dtype(dtype):
+            dtype = object
+
+        if fill_value is lib.no_default:
+            fill_value = np.nan
+
+        arr = np.empty(arr.n, dtype=dtype)
+        arr.fill(fill_value)
+        return arr
```

Comment:
@jbrockmendel do you know if we have similar functionality already existing? 
(There are some similar pieces of code in `internals/concat.py`, but those are quite entangled with the join units / blocks)

Reply:
isn't this basically na_value_for_dtype? (then just filling it)

Reply:
Hmm, `na_value_for_dtype` is certainly related, but not sure it would directly help here. Because I need to handle the EA / datetime / timedelta case separately anyhow, and then for the last `else: ` block to create a numpy ndarray, it's basically only ever `np.nan` (unless overridden with the keyword)

Reply:
You're right both that there is a lot of very similar code and that it isnt in a form that is easy to reuse.

Parts of this function seem to do the same thing as _dtype_to_na_value.  Other parts look very similar to the `if self.is_na` portion of `JoinUnit.get_reindexed_values`.

FWIW instead of handling the DTA/TDA cases separately, id do them in the numpy block and then call construction.ensure_wrapped_if_datetimelike

Reply:
> FWIW instead of handling the DTA/TDA cases separately, id do them in the numpy block and then call construction.ensure_wrapped_if_datetimelike

Ah, that sounds as a good idea, that should simplify this somewhat (and then also makes more sense to re-use `na_value_for_dtype`). Thanks